mtp_api/apps/mtp_auth/validators.py

```diff
@@ -1,16 +1,31 @@
+from django.contrib.auth import get_user_model
 from oauth2_provider.oauth2_validators import OAuth2Validator
 
-from .models import ApplicationUserMapping
+from .models import ApplicationUserMapping, FailedLoginAttempt
 
 
-class ApplicationRequestValidator(OAuth2Validator):
+class LockedOut(Exception):
+    pass
+
 
+class ApplicationRequestValidator(OAuth2Validator):
     def validate_user(self, username, password, client, request, *args, **kwargs):
-        valid = super(ApplicationRequestValidator, self).validate_user(
-            username, password, client, request, *args, **kwargs
-        )
-        if valid and ApplicationUserMapping.objects.filter(
-                user=request.user, application=client).exists():
-            return True
+        user = get_user_model().objects.filter(username=username).first()
+        try:
+            if user and FailedLoginAttempt.is_locked_out(user, client):
+                raise LockedOut
+
+            valid = super().validate_user(
+                username, password, client, request, *args, **kwargs
+            )
+            if valid and ApplicationUserMapping.objects.filter(
+                    user=request.user, application=client).exists():
+                FailedLoginAttempt.delete_failed_attempts(user, client)
+                return True
+            elif user:
```

Comment:
unfortunately, i don't think the user's IP can be passed on to the api app from the client app as part the oauth2 token request, unless we do some hacking and thereby potentially depart from the oauth2 standard…

similarly, without departing from the oauth2 standard, we cannot actually report to the end user that they are locked out directly as part of the token request. it'd still be possible to have a separate api method to determine whether a lockout exists, but then we'd likely be revealing whether a username exists or not…